sql: allow multiple UDFs with in-exact type parameters of the same family

il9ue opened the pull request:
This patch makes it possible for udf to check (#94146 )
 exact type matching for udf parameters
 Added TypeList.MatchExact to apply udf parameter check only.

 Unit test for such family type parameter matching are provided

 Release note: None

il9ue committed:
tree: allow multiple UDFs with in-exact type parameters

This patch makes it possible for udf to check
 exact type matching for udf parameters
 Added TypeList.MatchExact to apply udf parameter check only.

 Unit test for such family type parameter matching are provided

 Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/95147)
<!-- Reviewable:end -->

jordanlewis commented:
cc @mgartner 

Thanks for the PR! @il9ue, please add tests for this functionality. You can add tests to this file: `pkg/sql/logictest/testdata/logic_test/udf `

To run that test run `./dev testlogic base --config=local --files=udf`

il9ue committed:
tree: allow multiple UDFs with in-exact type parameters
This patch makes it possible for udf to check
exact type matching for udf parameters
Added TypeList.MatchExact to apply udf parameter check only.

- applied review : use identical func instead of newly created
- added e2e test on pkg/sql/logictest/testdata/logic_test/udf

Release note: None

il9ue committed:
Merge remote-tracking branch 'upstream/master'

il9ue committed:
Merge remote-tracking branch 'upstream/master'

il9ue committed:
tree: allow multiple UDFs with in-exact type parameters

This patch makes it possible for udf to check
exact type matching for udf parameters
Added TypeList.MatchExact to apply udf parameter check only.

- applied review : use identical func instead of newly created
- added e2e test on pkg/sql/logictest/testdata/logic_test/udf

Release note: None

il9ue commented:
Let me prepare sorted patch (messed up)
